autotraining.lua

```diff
@@ -0,0 +1,280 @@
+        end
+        if isIgnNoble then
+            ignore_count = ignore_count +1
+            goto next_unit
+        end
+        table.insert(ret, unit)
+        ::next_unit::
+    end
+    return ret
+end
+
+function getTrainingSquads()
+    local squads = {}
+    for squad_id, _ in pairs(state.training_squads) do
+        local squad = df.squad.find(squad_id)
+        if squad then
+            table.insert(squads, squad)
+        else
+            -- setting to nil during iteration is permitted by lua
+            state.training_squads[squad_id] = nil
+        end
+    end
+    return squads
+end
+
+function getTrainingNeed(unit)
+    if unit == nil then return nil end
+    local needs =  unit.status.current_soul.personality.needs
+    for _, need in ipairs(needs) do
+        if need.id == MartialTraining then
+            return need
+        end
+    end
+    return nil
+end
+
+--######
+--Main
+--######
+
+-- Find all training squads
+-- Abort if no squads found
+function checkSquads()
+    local squads = {}
+    for _, squad in ipairs(getTrainingSquads()) do
+        if squad.entity_id == df.global.plotinfo.group_id then
+            local leader = squad.positions[0].occupant
+            if leader ~= -1 then
+                table.insert(squads,squad)
+            end
+        end
+    end
+
+    if #squads == 0 then
+        return nil
+    end
+
+    return squads
+end
+
+function addTraining(unit,good_squads)
+    if unit.military.squad_id ~= -1 then
+        for _, squad in ipairs(good_squads) do
+            if unit.military.squad_id == squad.id then
+                return true
+            end
+        end
+        return false
+    end
+    for _, squad in ipairs(good_squads) do
+        for i=1,9,1   do
+            if squad.positions[i].occupant  == -1 then
+                return dfhack.military.addToSquad(unit.id,squad.id,i)
+            end
+        end
+    end
+
+    return false
+end
+
+function removeAll()
+    if state.training_squads == nil then return end
+    for _, squad in ipairs(getTrainingSquads()) do
+        for i=1,9,1 do
+            local hf = df.historical_figure.find(squad.positions[i].occupant)
+            if hf ~= nil then
+                dfhack.military.removeFromSquad(hf.unit_id)
+            end
+        end
+    end
+end
+
+
+function check()
+    local squads = checkSquads()
+    local intraining_count = 0
+    local inque_count = 0
+    if squads == nil then return end
+    for _,squad in ipairs(squads) do
+        for i=1,9,1   do
+            if squad.positions[i].occupant  ~= -1 then
+                local hf = df.historical_figure.find(squad.positions[i].occupant)
+                if hf ~= nil then
+                    local unit = df.unit.find(hf.unit_id)
+                    local training_need = getTrainingNeed(unit)
+                    if not training_need or training_need.focus_level >= state.threshold then
+                        dfhack.military.removeFromSquad(unit)
```

Comment:
```suggestion
                        dfhack.military.removeFromSquad(unit.id)
```
Please test your code, this is an outright argument error.